gimie/io.py

```diff
@@ -4,102 +4,101 @@
+    """Provides read-only access to remote data via a file-like interface.
 
     Parameters
     ----------
-    name: the name of the resource, typically the filename.
-    url: the URL where the resource. can be downladed from.
-    headers: optional headers to pass to the request.
+    url:
+        The URL where the resource. can be downladed from.
+    headers:
+        Optional headers to pass to the request.
 
     Examples
     --------
-    >>> from gimie.io import RemoteResource
     >>> url = "https://raw.githubusercontent.com/SDSC-ORD/gimie/main/README.md"
-    >>> resource = RemoteResource("README.md", url)
+    >>> content = RemoteResource("README.md", url).open().read()
+    >>> assert isinstance(content, bytes)
     """
 
     def __init__(self, name: str, url: str, headers: Optional[dict] = None):
         self.name = name
         self.url = url
         self.headers = headers or {}
 
-    def open(self) -> io.BufferedReader:
+    def open(self) -> io.RawIOBase:
         resp = requests.get(
             self.url, headers=self.headers, stream=True
         ).iter_content(chunk_size=128)
-        return iterable_to_stream(resp)
+        return IterStream(resp)
 
 
-def iterable_to_stream(
-    iterable, buffer_size=io.DEFAULT_BUFFER_SIZE
-) -> io.BufferedReader:
-    """
-    Converts an iterable yielding bytestrings to a read-only input stream.
-    Lets you use an iterable (e.g. a generator) that yields bytestrings as a read-only
-    input stream.
+class IterStream(io.RawIOBase):
+    """Wraps an iterator under a like a file-like interface.
 
-    The stream implements Python 3's newer I/O API (available in Python 2's io module).
-    For efficiency, the stream is buffered.
+    Parameters
+    ----------
+    iterator:
+        An iterator yielding bytes.
 
-    credits: https://stackoverflow.com/a/20260030/8440675
+    Examples
+    --------
+    >>> stream = IterStream(iter([b"Hello ", b"World"]))
+    >>> stream.read()
+    b'Hello World'
     """
 
-    class IterStream(io.RawIOBase):
-        def __init__(self):
-            self.leftover = ""
-
-        def readable(self):
-            return True
-
-        def readinto(self, b):
-            try:
-                l = len(b)  # We're supposed to return at most this much
-                chunk = self.leftover or next(iterable)
-                output, self.leftover = chunk[:l], chunk[l:]
-                b[: len(output)] = output
-                return len(output)
-            except StopIteration:
-                return 0  # indicate EOF
-
-    return io.BufferedReader(IterStream(), buffer_size=buffer_size)
+    def __init__(self, iterator: Iterator[bytes]):
+        self.leftover = b""
+        self.iterator = iterator
+
+    def readable(self):
+        return True
+
+    def readinto(self, b):
+        try:
+            l = len(b)  # We're supposed to return at most this much
+            chunk = self.leftover or next(self.iterator)
```

Comment:
I believe the stream would reach an early end-of-file if the iterator yields an empty byte string, even if the iterator is not exhausted. I would document that this should be an iterator of non-empty byte sequences, and check that this is the case. Alternatively, you could loop until a non-empty chunk is yielded.

Reply:
Thanks, good catch! :) I opted to skip empty chunks with a while statement in a36af1dae01dcf4db81e63f14735519393da8a84